### PR TITLE
support cypher parameter

### PR DESCRIPTION
**Pull request by czpmango**
Related pr:
 - nebula-go  [https://github.com/vesoft-inc/nebula-go/pull/130](https://github.com/vesoft-inc/nebula-go/pull/130)
 - nebula-console  [https://github.com/vesoft-inc/nebula-console/pull/127](https://github.com/vesoft-inc/nebula-console/pull/127)
 - nebula-python [https://github.com/vesoft-inc/nebula-python/pull/147](https://github.com/vesoft-inc/nebula-python/pull/147)

### CODE REVIEW

File: src/common/expression/ParameterExpression.h [hunk 1]
```diff
@@ -0,0 +1,50 @@
+/* Copyright (c) 2021 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#ifndef COMMON_EXPRESSION_PARAMETEREXPRESSION_H_
```
Comment (jievince) [inline src/common/expression/ParameterExpression.h:7]: use `#pragma once`  instead. https://confluence.nebula-graph.io/pages/viewpage.action?pageId=10723532

File: src/graph/visitor/FoldConstantExprVisitor.cpp [hunk 2]
```diff
@@ -275,6 +275,11 @@ void FoldConstantExprVisitor::visit(DestPropertyExpression *expr) {
   canBeFolded_ = false;
 }
 
+void FoldConstantExprVisitor::visit(ParameterExpression *expr) {
+  UNUSED(expr);
+  canBeFolded_ = false;
```
Comment (jievince) [inline src/graph/visitor/FoldConstantExprVisitor.cpp:280]: Could be pushed down?

File: src/common/expression/ParameterExpression.cpp [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+const Value& ParameterExpression::eval(ExpressionContext& ectx) { return ectx.getVar(name_); }
+
+std::string ParameterExpression::toString() const { return name_; }
+
+bool ParameterExpression::operator==(const Expression& rhs) const {
+  return kind_ == rhs.kind() && name_ == rhs.toString();
+}
+
+void ParameterExpression::writeTo(Encoder& encoder) const {
+  encoder << kind_;
```
Comment (jievince) [inline src/common/expression/ParameterExpression.cpp:22]: Add encode/decode test.

File: src/graph/visitor/FoldConstantExprVisitor.cpp (hunk 2 above)
Comment (czpmango) [reply src/graph/visitor/FoldConstantExprVisitor.cpp:280]: Not yet.

File: src/common/expression/ParameterExpression.cpp (hunk 3 above)
Comment (czpmango) [reply src/common/expression/ParameterExpression.cpp:22]: ACK.

File: src/common/expression/ParameterExpression.h (hunk 1 above)
Comment (czpmango) [reply src/common/expression/ParameterExpression.h:7]: ACK.

File: src/graph/context/QueryContext.cpp [hunk 4]
```diff
@@ -30,6 +30,14 @@ void QueryContext::init() {
...
   ep_ = std::make_unique<ExecutionPlan>();
   ectx_ = std::make_unique<ExecutionContext>();
+  // copy parameterMap into ExecutionContext
+  std::unordered_map<std::string, Value> paramMap;
+  if (rctx_) {
+    paramMap = rctx_->parameterMap();
+  }
+  for (auto item : paramMap) {
+    ectx_->setValue(item.first, std::move(item.second));
+  }
```
Comment (yixinglu) [inline src/graph/context/QueryContext.cpp:40]:
```c++
  if (rctx_) {
    for (auto item : rctx_->parameterMap()) {
      ectx_->setValue(item.first, std::move(item.second));
    }
  }
```

File: src/graph/context/QueryContext.h [hunk 5]
```diff
@@ -97,6 +97,8 @@ class QueryContext {
 
   bool isKilled() const { return killed_.load(); }
 
+  bool existParameter(const std::string param) const { return ectx_->exist(param); }
```
Comment (yixinglu) [inline src/graph/context/QueryContext.h:100]: const std::string &param

Comment (czpmango) [reply src/graph/context/QueryContext.h:100]: ACK.

File: src/graph/context/QueryContext.cpp (hunk 4 above)
Comment (czpmango) [reply src/graph/context/QueryContext.cpp:40]: ACK.

File: src/graph/service/RequestContext.h [hunk 6]
```diff
@@ -68,6 +68,12 @@ class RequestContext final : public cpp::NonCopyable, public cpp::NonMovable {
 
   GraphSessionManager* sessionMgr() const { return sessionMgr_; }
 
+  void setParameterMap(std::unordered_map<std::string, Value> parameterMap) {
+    parameterMap_ = parameterMap;
+  }
+
+  const std::unordered_map<std::string, Value> parameterMap() const { return parameterMap_; }
```
[the remark this answers is not in the document]
Comment (czpmango) [reply src/graph/service/RequestContext.h:75]: ACK.

File: src/graph/service/RequestContext.h [hunk 7]
```diff
@@ -68,6 +68,12 @@ class RequestContext final : public cpp::NonCopyable, public cpp::NonMovable {
 
   GraphSessionManager* sessionMgr() const { return sessionMgr_; }
 
+  void setParameterMap(std::unordered_map<std::string, Value> parameterMap) {
+    parameterMap_ = parameterMap;
```
[the remark this answers is not in the document]
Comment (czpmango) [reply src/graph/service/RequestContext.h:72]: ACK.

File: src/common/expression/ParameterExpression.h [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+// The ParameterExpression use for parameterized statement
+namespace nebula {
+
+class ParameterExpression : public Expression {
+ public:
+  ParameterExpression& operator=(const ParameterExpression& rhs) = delete;
+  ParameterExpression& operator=(ParameterExpression&&) = delete;
+
+  static ParameterExpression* make(ObjectPool* pool, const std::string& name = "") {
+    DCHECK(!!pool);
```
[the remark this answers is not in the document]
Comment (czpmango) [reply src/common/expression/ParameterExpression.h:20]: ACK.

File: src/common/expression/ParameterExpression.cpp [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+  encoder << name_;
+}
```
Comment (yixinglu) [inline src/common/expression/ParameterExpression.cpp:24]: Will this expression be passed to the storage layer?

File: tests/tck/features/yield/parameter.feature [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    Then the result should be, in any order:
+      | ival | bval  | sval           | fval |
+      | 2    | false | "Tim Duncanef" | 4.1  |
+
+  Scenario: match with parameters
+    When executing query:
+      """
+      MATCH (v:player)-[:like]->(n) where id(v)==$p3 and n.age>$p1+29
+      RETURN n.name AS dst
+      """
```
Comment (yixinglu) [inline tests/tck/features/yield/parameter.feature:26]: It seems that a lot of test cases need to be added, such as:
* $var = GO ..
* LOOKUP ON .. WHERE player.name == $p
* GO with pipe
* $var and $parameter mixed use
* ...

**Review by yixinglu — COMMENTED** [review]
Does this PR not take into account the implementation of parameter binding after the execution plan has been optimized?

File: src/graph/service/GraphService.cpp [hunk 11]
```diff
@@ -153,13 +153,64 @@ folly::Future<ExecutionResponse> GraphService::future_execute(int64_t sessionId,
   return future;
 }
 
+folly::Future<ExecutionResponse> GraphService::future_executeWithParameter(
+    int64_t sessionId,
+    const std::string& query,
+    const std::unordered_map<std::string, Value>& parameterMap) {
+  auto ctx = std::make_unique<RequestContext<ExecutionResponse>>();
```
[the remark this answers is not in the document]
Comment (czpmango) [reply src/graph/service/GraphService.cpp:160]: ACK.

File: src/graph/service/GraphService.cpp [hunk 12]
```diff
@@ -153,13 +153,64 @@ folly::Future<ExecutionResponse> GraphService::future_execute(int64_t sessionId,
...
   return folly::toJson(respJsonObj);
 }
 
+folly::Future<std::string> GraphService::future_executeJsonWithParameter(
+    int64_t sessionId,
+    const std::string& query,
+    const std::unordered_map<std::string, Value>& parameterMap) {
+  auto rawResp = future_executeWithParameter(sessionId, query, parameterMap).get();
+  auto respJsonObj = rawResp.toJson();
+  return folly::toJson(respJsonObj);
```
[the remark this answers is not in the document]
Comment (czpmango) [reply src/graph/service/GraphService.cpp:211]: ACK.

File: tests/Makefile [hunk 13]
```diff
@@ -22,7 +22,8 @@ install-deps:
 	pip3 install --user -r $(CURR_DIR)/requirements.txt -i $(PYPI_MIRROR)
 
 install-nebula-py: install-deps
-	git clone --branch master https://github.com/vesoft-inc/nebula-python $(CURR_DIR)/nebula-python
+    # TODO: just for test  (czp)
+	git clone --branch param-cmd https://github.com/czpmango/nebula-python $(CURR_DIR)/nebula-python
```
[the remark this answers is not in the document]
Comment (czpmango) [reply tests/Makefile:26]: Ok.

File: src/common/expression/ParameterExpression.cpp (hunk 9 above)
Comment (czpmango) [reply src/common/expression/ParameterExpression.cpp:24]: Not yet.ParameterExpression relies on graph::ExecutionContext::valueMap_.

File: tests/tck/features/yield/parameter.feature (hunk 10 above)
Comment (czpmango) [reply tests/tck/features/yield/parameter.feature:26]: In progress.

**Comment by czpmango** [comment]
> Does this PR not take into account the implementation of parameter binding after the execution plan has been optimized?

Parameter push-down is currently not supported.
refer [https://github.com/vesoft-inc/nebula/issues/2593](https://github.com/vesoft-inc/nebula/issues/2593) for more detail.

File: src/graph/service/GraphService.cpp [hunk 14]
```diff
@@ -116,6 +116,14 @@ void GraphService::signout(int64_t sessionId) {
 
 folly::Future<ExecutionResponse> GraphService::future_execute(int64_t sessionId,
                                                               const std::string& query) {
+  std::unordered_map<std::string, Value> params;
+  return future_executeWithParameter(sessionId, query, std::move(params)).get();
```
Comment (jievince) [inline src/graph/service/GraphService.cpp:120]: return future_executeWithParameter(sessionId, query, std::unordered_map<std::string, Value>{}).get();

File: src/graph/service/GraphService.cpp [hunk 15]
```diff
@@ -155,7 +164,17 @@ folly::Future<ExecutionResponse> GraphService::future_execute(int64_t sessionId,
 
 folly::Future<std::string> GraphService::future_executeJson(int64_t sessionId,
                                                             const std::string& query) {
-  auto rawResp = future_execute(sessionId, query).get();
+  std::unordered_map<std::string, Value> params;
+  auto rawResp = future_executeWithParameter(sessionId, query, std::move(params)).get();
```
Comment (jievince) [inline src/graph/service/GraphService.cpp:168]: Ditto

Comment (czpmango) [reply src/graph/service/GraphService.cpp]: Fixed.

File: src/graph/service/GraphService.cpp (hunk 14 above)
Comment (czpmango) [reply src/graph/service/GraphService.cpp]: ACK.

File: tests/tck/features/yield/parameter.feature [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+    Then the result should be, in any order:
+      | dst             |
+      | "Manu Ginobili" |
+      | "Tony Parker"   |
+
+  Scenario: var with parameters
+    When executing query:
+      """
+      $p1=GO FROM "Tim Duncan" OVER like WHERE like.likeness>$p1 yield like._dst as dst;
+      GO FROM $p1.dst OVER like YIELD DISTINCT $$.player.name AS name
```
Comment (CPWstatic) [inline tests/tck/features/yield/parameter.feature:35]:
```
 $p1=GO FROM "Tim Duncan" OVER like WHERE like.likeness>$p1 yield like._dst as dst;
 GO FROM $p1.dst OVER like Where like.age > $p1 YIELD DISTINCT $$.player.name AS name.
```
What is the behavior of this test case?
We need more test cases.

File: src/common/expression/ParameterExpression.h [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+ */
+
+#pragma once
+
+#include "common/expression/Expression.h"
+
+// The ParameterExpression use for parameterized statement
+namespace nebula {
+
+class ParameterExpression : public Expression {
```
Comment (CPWstatic) [inline src/common/expression/ParameterExpression.h:14]: I doubt if we need this one.

File: tests/tck/features/yield/parameter.feature (hunk 16 above)
Comment (czpmango) [reply tests/tck/features/yield/parameter.feature:35]: `$p1` started as a parameter and was rewritten as a variable, just consistent with what we talked about earlier.
---

Such as what test cases？

File: src/common/expression/ParameterExpression.h (hunk 17 above)
Comment (czpmango) [reply src/common/expression/ParameterExpression.h:14]: Is need. 
Such as lookupValidators or some visitor or scenes that require expression rewriting.

Some Example :
  - rewrite ParameterExpression to ConstantExpression
  - EvaluableExprVisitor determines whether expressions are evaluable by their specific types. Code likes that:
  ```
void visit(VariableExpression *) override { isEvaluable_ = false; }
```
  - If `ParameterExpression` and `VariableExpression` are unified, then we must use if-else whenever we need to distinguish them, I don't think this is a good code design pattern.

File: tests/tck/features/yield/parameter.feature (hunk 16 above)
Comment (CPWstatic) [reply tests/tck/features/yield/parameter.feature:35]: Where did you do the rewriting?

Comment (czpmango) [reply tests/tck/features/yield/parameter.feature:35]: Expression rewriting?

Comment (czpmango) [reply tests/tck/features/yield/parameter.feature:35]: The implementation layer is the same as `VariableExpression`, it is based on `valueMap_`, so there is no need to do rewriting. `valueMap_` is automatically written to at runtime.

**Review by jievince — APPROVED** [review]
LGTM now

File: src/graph/service/GraphService.cpp [hunk 18]
```diff
@@ -116,6 +116,14 @@ void GraphService::signout(int64_t sessionId) {
 
 folly::Future<ExecutionResponse> GraphService::future_execute(int64_t sessionId,
                                                               const std::string& query) {
+  return future_executeWithParameter(sessionId, query, std::unordered_map<std::string, Value>{})
+      .get();
```
Comment (yixinglu) [inline src/graph/service/GraphService.cpp:120]: Why need to block this async call?

File: src/graph/service/GraphService.cpp [hunk 19]
```diff
@@ -155,7 +164,17 @@ folly::Future<ExecutionResponse> GraphService::future_execute(int64_t sessionId,
 
 folly::Future<std::string> GraphService::future_executeJson(int64_t sessionId,
                                                             const std::string& query) {
-  auto rawResp = future_execute(sessionId, query).get();
+  auto rawResp =
+      future_executeWithParameter(sessionId, query, std::unordered_map<std::string, Value>{}).get();
+  auto respJsonObj = rawResp.toJson();
+  return folly::toJson(respJsonObj);
```
Comment (yixinglu) [inline src/graph/service/GraphService.cpp:170]: DITTO;

```cpp
return future_executeWithParameter(sessionId, query, std::unordered_map<std::string, Value>{})
    .then([](auto &resp) {
      return folly::toJson(resp.toJson());
    });
```

File: src/graph/service/GraphService.cpp (hunk 18 above)
Comment (czpmango) [reply src/graph/service/GraphService.cpp]: Fixed.

File: src/graph/service/GraphService.cpp (hunk 19 above)
Comment (czpmango) [reply src/graph/service/GraphService.cpp]: Fixed.

File: src/graph/service/GraphService.cpp (hunk 18 above)
Comment (czpmango) [reply src/graph/service/GraphService.cpp]: Fixed.

**Review by jievince — APPROVED** [review]

**Review by yixinglu — DISMISSED** [review]

**Review by yixinglu — APPROVED** [review]

File: tests/tck/features/yield/parameter.feature (hunk 16 above)
Comment (czpmango) [reply tests/tck/features/yield/parameter.feature:49]: > ```
>  $p1=GO FROM "Tim Duncan" OVER like WHERE like.likeness>$p1 yield like._dst as dst;
>  GO FROM $p1.dst OVER like Where like.age > $p1 YIELD DISTINCT $$.player.name AS name.
> ```
> 
> What is the behavior of this test case? We need more test cases.

Variables and parameters with the same name are not supported for the time being, just consistent with our offline discussion.

**Pull request closed without merging**